src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -156,6 +222,673 @@ MultiRouterPlanCreate(Query *originalQuery, Query *query,
+	{
+		ereport(DEBUG2, (errmsg("could not generate task for target shardId: %ld",
+								shardId),
+						 errdetail("Insert query hits %d placements, Select query "
+								   "hits %d placements and only %d of those placements match.",
+								   list_length(insertShardPlacementList),
+								   list_length(selectPlacementList),
+								   list_length(intersectedPlacementList))));
+
+		return NULL;
+	}
+
+	/* this is required for correct deparsing of the query */
+	ReorderInsertSelectTargetLists(copiedQuery, copiedInsertRte, copiedSubqueryRte);
+
+	/* set the upsert flag */
+	if (originalQuery->onConflict != NULL)
+	{
+		upsertQuery = true;
+	}
+
+	/* setting an alias simplifies deparsing of RETURNING */
+	if (copiedInsertRte->alias == NULL)
+	{
+		Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
+		copiedInsertRte->alias = alias;
+	}
+
+	/* and generate the full query string */
+	deparse_shard_query(copiedQuery, distributedTableId, shardInterval->shardId,
+						queryString);
+	ereport(DEBUG4, (errmsg("distributed statement: %s", queryString->data)));
+
+	modifyTask = CreateBasicTask(jobId, taskIdIndex, MODIFY_TASK, queryString->data);
+	modifyTask->dependedTaskList = NULL;
+	modifyTask->anchorShardId = shardId;
+	modifyTask->taskPlacementList = insertShardPlacementList;
+	modifyTask->upsertQuery = upsertQuery;
+	modifyTask->selectShardList = selectShardList;
+
+	return modifyTask;
+}
+
+
+/*
+ * AddShardIntervalRangeToSelect adds the following range boundaries
+ * with the given subquery and shardInterval:
+ *
+ *    hashfunc(partitionColumn) >= $lower_bound AND
+ *    hashfunc(partitionColumn) <= $upper_bound
+ *
+ * The function expects and asserts that subquery's target list contains a partition
+ * column value.
+ */
+static void
+AddShardIntervalRangeToSelect(Query *subqery, ShardInterval *shardInterval)
+{
+	List *targetList = subqery->targetList;
+	ListCell *targetEntryCell = NULL;
+	Var *targetPartitionColumnVar = NULL;
+	Oid integer4GEoperatorId = InvalidOid;
+	Oid integer4LEoperatorId = InvalidOid;
+	TypeCacheEntry *typeEntry = NULL;
+	FuncExpr *hashFunctionExpr = NULL;
+	OpExpr *greaterThanAndEqualsBoundExpr = NULL;
+	OpExpr *lessThanAndEqualsBoundExpr = NULL;
+
+	/* iterate through the target entries */
+	foreach(targetEntryCell, targetList)
+	{
+		TargetEntry *targetEntry = lfirst(targetEntryCell);
+
+		if (IsPartitionColumnRecursive(targetEntry->expr, subqery) &&
+			IsA(targetEntry->expr, Var))
+		{
+			targetPartitionColumnVar = (Var *) targetEntry->expr;
+			break;
+		}
+	}
+
+	/* we should have found target partition column */
+	Assert(targetPartitionColumnVar != NULL);
+
+	integer4GEoperatorId = get_opfamily_member(INTEGER_BTREE_FAM_OID, INT4OID,
+											   INT4OID,
+											   BTGreaterEqualStrategyNumber);
+	integer4LEoperatorId = get_opfamily_member(INTEGER_BTREE_FAM_OID, INT4OID,
+											   INT4OID,
+											   BTLessEqualStrategyNumber);
+
```

Comment:
Maybe add an assert or error check for get_opfamily_member returning InvalidOid?